request_logger.go

```diff
@@ -0,0 +1,79 @@
+package xecho
+
+import (
+	"net/http"
+	"time"
+
+	"github.com/labstack/echo"
+	"github.com/sirupsen/logrus"
+)
+
+type TimeProvider func() time.Time
+
+func RequestLoggerMiddleware(timeFn TimeProvider) echo.MiddlewareFunc {
+	return func(next echo.HandlerFunc) echo.HandlerFunc {
+		return EchoHandler(func(c *Context) error { return RequestLogger(c, next, timeFn) })
+	}
+}
+
+func RequestLogger(c *Context, next echo.HandlerFunc, time TimeProvider) error {
+	before := time()
+	lrw := &statefulResponseWriter{ResponseWriter: c.Response().Writer}
+	c.Response().Writer = lrw
+	err := next(c)
+	after := time()
+	logger, ok := c.Logger().(*Logger)
+	if !ok {
+		c.Logger().Infof("[%s] %s %d", c.Request().Method, c.Path(), lrw.statusCode)
+		return err
+	}
+	logger.
+		WithFields(createMap(c, after.Sub(before), lrw, err)).
+		Infof("[%s] %s %d", c.Request().Method, c.Path(), lrw.statusCode)
+	return err
+}
+
+func createMap(c echo.Context, timeTaken time.Duration, lrw *statefulResponseWriter, err error) logrus.Fields {
+	r := c.Request()
+	fields := logrus.Fields{
+		"duration_ms": milliseconds(timeTaken),
```

Comment:
Here, you can just do `timeTaken.Milliseconds()`

Reply:
yeah I did that originally but this framework is in 1.12 where this does not exist (so I copied it from 1.13)

Reply:
If we converted it, it would be a good incentive for other to upgrade their projects when importing this 😉